perf/tools/setup_env.sh

```diff
@@ -41,91 +68,50 @@ if [ -f "$sysfs_path" ]; then
 else
-  echo "SKIPPED"
+  echo $SKIPPED_MSG
 fi
 
-echo -n "Isolate CPUs for benchmarking "
+desc_msg="Isolate CPUs for benchmarking"
 sysfs_path="/sys/devices/system/cpu/isolated"
-if [ -f "$sysfs_path" ]; then
-  isolcpus=$(cat "$sysfs_path")
-  [ -n "$isolcpus" ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'isolcpus' kernel parameter)"
-else
-  echo "SKIPPED"
-fi
+hint_msg="set 'isolcpus' kernel parameter"
+kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
 
-echo -n "Offload interrupts from the isolated CPUs "
+desc_msg="Offload interrupts from the isolated CPUs"
 sysfs_path="/proc/irq/default_smp_affinity"
-if [ -f "$sysfs_path" ]; then
-  smp_affinity=$(cat "$sysfs_path")
-  [ -n "$smp_affinity" ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'irqaffinity' kernel parameter)"
-else
-  echo "SKIPPED"
-fi
+hint_msg="set 'irqaffinity' kernel parameter"
+kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
 
-echo -n "Disable scheduling on single-task isolated CPUs "
+desc_msg="Disable scheduling on single-task isolated CPUs"
 sysfs_path="/sys/devices/system/cpu/nohz_full"
-if [ -f "$sysfs_path" ]; then
-  nohz=$(cat "$sysfs_path")
-  [ -n "$nohz" ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'nohz_full' kernel parameter)"
-else
-  echo "FAILED (hint: enable CONFIG_NO_HZ_FULL in a kernel config)"
-fi
+hint_msg="set 'nohz_full' kernel parameter"
+kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
 
-echo -n "Disable transparent_hugepage "
+desc_msg="Disable transparent huge pages"
 sysfs_path="/sys/kernel/mm/transparent_hugepage/enabled"
-if [ -f "$sysfs_path" ]; then
-  sh -c "echo never > $sysfs_path"
-  echo $SUCCESS_MSG
-else
-  echo $SKIPPED_MSG
-fi
+set_kernel_setting "$desc_msg" "$sysfs_path" "never"
 
-echo -n "Disable transparent_hugepage "
+desc_msg="Disable direct compaction of transparent huge pages "
 sysfs_path="/sys/kernel/mm/transparent_hugepage/defrag"
-if [ -f "$sysfs_path" ]; then
-  sh -c "echo never > $sysfs_path"
-  echo $SUCCESS_MSG
-else
-  echo $SKIPPED_MSG
-fi
+set_kernel_setting "$desc_msg" "$sysfs_path" "never"
 
 ###
 # System tuning.
 ###
 
-echo -n "Disable TurboBoost "
+desc_msg="Disable TurboBoost"
 if [ "$cpu_vendor" = "amd" ]; then
   sysfs_path="/sys/devices/system/cpu/cpufreq/boost"
-  if [ -f "$sysfs_path" ]; then
-    sh -c "echo 0 > $sysfs_path"
-    echo $SUCCESS_MSG
-  else
-    echo $SKIPPED_MSG
-  fi
+  value=0
 elif [ "$cpu_vendor" = "intel" ]; then
   sysfs_path="/sys/devices/system/cpu/intel_pstate/no_turbo"
-  if [ -f "$sysfs_path" ]; then
-    sh -c "echo 1 > $sysfs_path"
-    echo $SUCCESS_MSG
-  else
-    echo $SKIPPED_MSG
-  fi
+  value=1
 fi
+set_kernel_setting "$desc_msg" "$sysfs_path" "$value"
 
-echo -n "Stabilize the CPU frequency "
+desc_msg="Stabilize the CPU frequency"
 sysfs_path="/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor"
```

Comment:
The `-f "$filename"` doesn't work correctly with wildcard symbol:

I've checked that locally:

```sh
desc_msg="TEST MASK:"
sysfs_path="/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor"
kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "TESTMASK:"

desc_msg="TEST RAW:"
sysfs_path="/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"
kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "TESTRAW:"

# ...

TEST MASK: SKIPPED
TEST RAW: CHECKED
```

See [this run](https://github.com/tarantool/tarantool/actions/runs/9939553668/job/27454348529?pr=10219#step:7:112).

Reply:
Updated:

```diff
@@ -113,10 +114,14 @@ set_kernel_setting \
   "$sysfs_path" \
   "$value"
 
-set_kernel_setting \
-  "Stabilize the CPU frequency" \
-  "/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor" \
-  "performance"
+sysfs_path="/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor"
+for sysfs_path_cpu in $sysfs_path; do
+  cpu_n=$(echo "$sysfs_path_cpu" | cut -d'/' -f6)
+  set_kernel_setting \
+    "Stabilize the CPU frequency $sysfs_path_cpu" \
+    "$sysfs_path_cpu" \
+    "performance"
+done
 
 set_kernel_setting \
   "Workload-specific: drop disk caches" \
```

Reply:
Locally, it works like a charm, but we still have troubles in CI, see [this run](https://github.com/tarantool/tarantool/actions/runs/9944968058/job/27474134158#step:7:103).

Reply:
Shell will not unfold mask when path doesn't exist:

On my personal machine (Ubuntu 22.04):

```
$ stat /sys/devices/system/cpu/cpu1/cpufreq/scaling_governor
  File: /sys/devices/system/cpu/cpu1/cpufreq/scaling_governor
  Size: 4096            Blocks: 0          IO Block: 4096   regular file
Device: 16h/22d Inode: 24057       Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-07-04 15:15:37.957732630 +0300
Modify: 2024-07-16 11:09:29.189046197 +0300
Change: 2024-07-16 11:09:29.189046197 +0300
 Birth: -
$ for cpu in /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor; do echo $cpu; done
/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor
/sys/devices/system/cpu/cpu1/cpufreq/scaling_governor
/sys/devices/system/cpu/cpu2/cpufreq/scaling_governor
/sys/devices/system/cpu/cpu3/cpufreq/scaling_governor
/sys/devices/system/cpu/cpu4/cpufreq/scaling_governor
/sys/devices/system/cpu/cpu5/cpufreq/scaling_governor
/sys/devices/system/cpu/cpu6/cpufreq/scaling_governor
/sys/devices/system/cpu/cpu7/cpufreq/scaling_governor
$ 
```

On ghacts-tarantool-8-16-n5 (Ubuntu 20.04):

```
$ stat /sys/devices/system/cpu/cpu1/cpufreq/scaling_governor
stat: cannot stat '/sys/devices/system/cpu/cpu1/cpufreq/scaling_governor': No such file or directory
$ for cpu in /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor; do echo $cpu; done
/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor
```

So I've updated this check in the script:

```diff
--- a/perf/tools/setup_env.sh
+++ b/perf/tools/setup_env.sh
@@ -119,11 +119,11 @@ set_kernel_setting \
   "$sysfs_path" \
   "$value"
 
-sysfs_path="/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor"
-for sysfs_path_cpu in $sysfs_path; do
-  cpu_n=$(echo "$sysfs_path_cpu" | cut -d'/' -f6)
+ncpu=$(getconf _NPROCESSORS_ONLN)
+for cpu_id in $(seq 0 1 $((ncpu-1))); do
+  sysfs_path_cpu="/sys/devices/system/cpu/cpu$cpu_id/cpufreq/scaling_governor"
   set_kernel_setting \
-    "Stabilize the frequency of CPU $cpu_n" \
+    "Stabilize the frequency of CPU $cpu_id" \
     "$sysfs_path_cpu" \
     "performance"
 done
```

However, on the aforementioned machine (ghacts-tarantool-8-16-n5) check is still skipped. The reason is missed SYSFS settings `scaling_governor` for CPUs. As I got it right, it is due to absence of CPU scaling driver [1]. I continue an investigation.

1. https://docs.kernel.org/admin-guide/pm/cpufreq.html

Reply:
There is no cpufreq driver on ghacts-tarantool-8-16-n5:
```
buntu@ghacts-tarantool-8-16-n5:~$ cpufreq-info
cpufrequtils 008: cpufreq-info (C) Dominik Brodowski 2004-2009
Report errors and bugs to cpufreq@vger.kernel.org, please.
analyzing CPU 0:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
analyzing CPU 1:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
analyzing CPU 2:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
analyzing CPU 3:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
analyzing CPU 4:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
analyzing CPU 5:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
analyzing CPU 6:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
analyzing CPU 7:
  no or unknown cpufreq driver is active on this CPU
  maximum transition latency: 4294.55 ms.
```

Reply:
Got it!
Thanks for the investigation!